## Feature Selection

In [1]:
import pandas as pd
import numpy as np


In [2]:
df=pd.read_csv("/content/employee_interview.csv")

df.head()

,Age,Years_of_Experience,Education_Level,Certifications_Count,Previous_Employers,Tech_Skill_Score,Communication_Score,Problem_Solving_Score,Leadership_Score,Teamwork_Score,...,Current_CTC,Expected_CTC,Notice_Period_Days,Job_Changes_Count,Projects_Completed,Hackathons_Participated,Client_Interaction_Score,Python_Proficiency,SQL_Proficiency,Target
0,50.0,4.0,2.0,1.0,3.0,86.0,70.0,61.0,98.0,48.0,...,4.19,16.70,31.0,0.0,6.0,2.0,91.0,42.0,58.0,1
1,60.0,0.0,1.0,5.0,0.0,63.0,81.0,84.0,34.0,72.0,...,10.38,19.86,35.0,3.0,0.0,0.0,69.0,0.0,79.0,1
2,18.0,12.0,1.0,3.0,3.0,81.0,57.0,54.0,37.0,85.0,...,7.63,1.00,23.0,2.0,1.0,0.0,44.0,66.0,65.0,0
3,19.0,2.0,5.0,1.0,4.0,63.0,74.0,100.0,85.0,76.0,...,2.92,1.00,54.0,2.0,14.0,5.0,89.0,100.0,69.0,0
4,60.0,0.0,2.0,-0.0,0.0,77.0,84.0,53.0,89.0,39.0,...,18.53,12.03,32.0,5.0,8.0,4.0,71.0,23.0,94.0,0


In [3]:
df.columns


Index(['Age', 'Years_of_Experience', 'Education_Level', 'Certifications_Count',
       'Previous_Employers', 'Tech_Skill_Score', 'Communication_Score',
       'Problem_Solving_Score', 'Leadership_Score', 'Teamwork_Score',
       'Work_Life_Balance_Score', 'Current_CTC', 'Expected_CTC',
       'Notice_Period_Days', 'Job_Changes_Count', 'Projects_Completed',
       'Hackathons_Participated', 'Client_Interaction_Score',
       'Python_Proficiency', 'SQL_Proficiency', 'Target'],
      dtype='object')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1000 non-null   float64
 1   Years_of_Experience       1000 non-null   float64
 2   Education_Level           1000 non-null   float64
 3   Certifications_Count      1000 non-null   float64
 4   Previous_Employers        1000 non-null   float64
 5   Tech_Skill_Score          1000 non-null   float64
 6   Communication_Score       1000 non-null   float64
 7   Problem_Solving_Score     1000 non-null   float64
 8   Leadership_Score          1000 non-null   float64
 9   Teamwork_Score            1000 non-null   float64
 10  Work_Life_Balance_Score   1000 non-null   float64
 11  Current_CTC               1000 non-null   float64
 12  Expected_CTC              1000 non-null   float64
 13  Notice_Period_Days        1000 non-null   float64
 14  Job_Chang

## Feature selection types

## 1. Filter Methods

### 1.1 Correlation:-  when we have continuous target variable

In [5]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

X =df.drop(['Target','Expected_CTC'],axis=1)
y=df['Expected_CTC']

# train test split
X_train,X_test,y_train,y_test=train_test_split(X,y, test_size=0.2, random_state=42)

# Initialize and train the multiple Regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

print('multiple Linear Regression Accuracy: ', round(r2_score(y_test, y_pred), 3))



multiple Linear Regression Accuracy:  0.792


In [6]:
# multiple linear regression assumes that there should not be much correlation between any two features ie. not multicollinear

for i in df.columns:
  for index, value in df.corr()[i].to_dict().items():
    if abs(value)>0.53 and abs(value)!=1: # threshould value >53% correlation it will return
      print(i, ',',index,round(value,2))

Certifications_Count , Projects_Completed -0.54
Previous_Employers , Job_Changes_Count -0.61
Expected_CTC , Client_Interaction_Score 0.57
Job_Changes_Count , Previous_Employers -0.61
Projects_Completed , Certifications_Count -0.54
Client_Interaction_Score , Expected_CTC 0.57


In [8]:
df.corr()[i].to_dict().items()

dict_items([('Age', -0.023807087957577574), ('Years_of_Experience', 0.03938446194522964), ('Education_Level', -0.2107219453538424), ('Certifications_Count', -0.02511819996836579), ('Previous_Employers', 0.12108848232857843), ('Tech_Skill_Score', -0.0681208280918773), ('Communication_Score', 0.020131227781054604), ('Problem_Solving_Score', 0.19433601098760772), ('Leadership_Score', 0.019020072515195066), ('Teamwork_Score', 0.2510529874860649), ('Work_Life_Balance_Score', 0.024050858446440202), ('Current_CTC', 0.2581188871830771), ('Expected_CTC', -0.02809872525264803), ('Notice_Period_Days', 0.04736410979002457), ('Job_Changes_Count', -0.422767695646025), ('Projects_Completed', -0.31829765183025327), ('Hackathons_Participated', -0.19559243023063405), ('Client_Interaction_Score', 0.2817789875393609), ('Python_Proficiency', 0.07097696482694581), ('SQL_Proficiency', -0.011643235247962671), ('Target', 1.0)])

In [9]:
i

'Target'

In [10]:
# now we will see how much corelation for each column with our target column
df.corr()['Expected_CTC'].sort_values()

,Expected_CTC
Projects_Completed,-0.319165
Problem_Solving_Score,-0.314635
Hackathons_Participated,-0.303393
Current_CTC,-0.219592
Python_Proficiency,-0.190006
Teamwork_Score,-0.090790
Tech_Skill_Score,-0.041756
Job_Changes_Count,-0.028765
Target,-0.028099
Work_Life_Balance_Score,-0.011123


In [12]:
# removing multicollinear features
X=df.drop(['Target','Expected_CTC','Previous_Employers','Certifications_Count'],axis=1)
y=df['Expected_CTC']

# train test split
X_train,X_test,y_train,y_test=train_test_split(X,y, test_size=0.2, random_state=42)

# Initialize and tarain the multiple regresssion model
model=LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

print('multiple Linear Regression Accuracy: ', round(r2_score(y_test, y_pred), 3))

multiple Linear Regression Accuracy:  0.792


-  even though we removed two features , that means those two features had no relation with the target variable, so those are irrelavant features  still we got 0.792

### 1.2 Chi-Square

In [13]:
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

# loading titanic dataset
df=sns.load_dataset('titanic')
df.head()


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [14]:
cat_df=df[['sex','embarked','class','alone','sibsp','parch','fare']]
cat_df.head()

,sex,embarked,class,alone,sibsp,parch,fare
0,male,S,Third,False,1,0,7.2500
1,female,C,First,False,1,0,71.2833
2,female,S,Third,True,0,0,7.9250
3,female,S,First,False,1,0,53.1000
4,male,S,Third,True,0,0,8.0500


In [15]:
# for categorical values which are not numeric we need to encode it

from sklearn.preprocessing import LabelEncoder

for col in ['sex','embarked','class','alone']:
  encoder=LabelEncoder()
  cat_df[col]=encoder.fit_transform(cat_df[col])

cat_df.head()


,sex,embarked,class,alone,sibsp,parch,fare
0,1,2,2,0,1,0,7.2500
1,0,0,0,0,1,0,71.2833
2,0,2,2,1,0,0,7.9250
3,0,2,0,0,1,0,53.1000
4,1,2,2,1,0,0,8.0500


In [16]:
# now we perform
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

X=cat_df
y=df['survived']

# train test split
X_train,x_test,y_train,y_test=train_test_split(X,y, test_size=0.2, random_state=42)

# initialize and train the logiestic Regression model
model=SVC(random_state=42)
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(x_test)

print('SVC Accuracy: ', round(accuracy_score(y_test, y_pred), 3))

SVC Accuracy:  0.654


In [24]:
#
from sklearn.feature_selection import chi2,SelectKBest, chi2
chi2_selector=SelectKBest(score_func=chi2, k=3) # k is top features
chi2_selector.fit(X[['sex','embarked','class','alone','sibsp','parch']],y)

# show selected features
chi2_scores=pd.Series(chi2_selector.scores_, index=['sex','embarked','class','alone','sibsp','parch'])
print("Chi2 Top Features: \n",chi2_scores.sort_values(ascending=False))


Chi2 Top Features: 
 sex         92.702447
class       54.465866
alone       14.640793
parch       10.097499
embarked     9.755456
sibsp        2.581865
dtype: float64


In [26]:
X_chi2=chi2_selector.fit_transform(X[['sex','embarked','class','alone','sibsp','parch']],y)
X_chi2
#

array([[1, 2, 0],
       [0, 0, 0],
       [0, 2, 1],
       ...,
       [0, 2, 0],
       [1, 0, 1],
       [1, 2, 1]])

In [28]:
X_train.shape

(712, 3)

In [27]:
# train test split
X_train,x_test,y_train,y_test=train_test_split(X_chi2,y, test_size=0.2, random_state=42)

model=SVC(random_state=42)
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(x_test)

print('SVC Accuracy: ', round(accuracy_score(y_test, y_pred), 3))
#

SVC Accuracy:  0.799


### 1.3 ANOVA

In [29]:
from sklearn.feature_selection import f_classif

selector=SelectKBest(score_func=f_classif, k=2)
X_anova=selector.fit_transform(X[['sex','embarked','class','alone','sibsp','parch']],y)

print("Original number of features: ", X.shape)
print("Reduced number of features: " ,X_anova.shape)

Original number of features:  (891, 7)
Reduced number of features:  (891, 2)


In [30]:
# train test split
X_train,x_test,y_train,y_test=train_test_split(X_anova,y, test_size=0.2, random_state=42)

# initialize and train the support model
model=SVC()
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(x_test)

print('SVC Accuracy: ', round(accuracy_score(y_test, y_pred), 3))

SVC Accuracy:  0.765


## 2. Wrapper Method

In [31]:
from sklearn.linear_model import LogisticRegression
from sklearn.feature_selection import RFE  # Recursive Feature Elimination

model=LogisticRegression()
rfe=RFE(model, n_features_to_select=2)
rfe.fit(X,y)

rfe_selected=pd.Series(rfe.support_, index=X.columns)
print("RFE Selected Features: \n", rfe_selected[rfe_selected==True].index.tolist())



RFE Selected Features: 
 ['sex', 'class']


In [32]:

X_train,x_test,y_train,y_test=train_test_split(X[['sex','class']],y, test_size=0.2, random_state=42)

model=LogisticRegression()
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(x_test)

print('Logistic Regression Accuracy: ', round(accuracy_score(y_test, y_pred), 3))

Logistic Regression Accuracy:  0.782


## 3. Embedded Method

- here we don't select any feature , it automatically selected by the model

In [33]:
from sklearn.ensemble import RandomForestClassifier

rf=RandomForestClassifier()
rf.fit(X,y)

rf_importance=pd.Series(rf.feature_importances_, index=X.columns)
print("Top Features from Random Forest : \n", rf_importance.sort_values(ascending=False).head())

Top Features from Random Forest : 
 fare     0.417310
sex      0.310475
class    0.100039
sibsp    0.054816
parch    0.051573
dtype: float64


In [35]:
X_train,X_test,y_train,y_test=train_test_split(X[['fare','sex','class']],y, test_size=0.2, random_state=42)

model=RandomForestClassifier(random_state=42)
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

print('Random Forest Accuracy: ', round(accuracy_score(y_test, y_pred), 3))

Random Forest Accuracy:  0.816


| **Method**                             | **Type**    | **Use When...**                                                                 | **Best For**                                   |
|----------------------------------------|-------------|----------------------------------------------------------------------------------|------------------------------------------------|
| **Correlation**                            | Filter      | You want to remove multicollinearity or quickly assess relationships            | Any problem (classification/regression)        |
| **Chi-Square**                             | Filter      | Target is categorical, and features are non-negative and categorical/discrete   | Classification                                 |
| **ANOVA F-test**                           | Filter      | Target is categorical, and features are continuous                              | Classification                                 |
| **RFE (Recursive Feature Elimination)**    | Wrapper     | You want high accuracy, don't mind slower training                              | Small to medium datasets                       |
| **Tree-based (Random Forest, XGBoost)**   | Embedded    | You want fast, accurate feature ranking                                         | Any problem (auto-handles nonlinearity)        |
